# 03 Contribution extraction (SHAP / EBM)

Extracts global feature importance and local contributions for the fixed test
instances (INSTANCE_IDS) for XGBoost and EBM using option 2 (Poisson log).

* XGBoost: SHAP TreeExplainer, contributions in log space (additive)
* EBM: interpret explain_local, terms also in log space

All outputs are saved as JSON in explanations/ and read by notebooks 04 to 06.

In [1]:
from __future__ import annotations

import sys, json
from pathlib import Path

sys.path.insert(0, str(Path.cwd().parent))

import joblib
import numpy as np
import pandas as pd

from utils import (
    INSTANCE_IDS,
    EXPLANATIONS_DIR, MODELS_DIR, RESULTS_DIR,
)
from utils.data import load_train_test
from utils.explanations import (
    build_global, build_local, save_explanation,
    build_global_curves, save_global_curves,
)

LOSS_KEY = "poisson_log"

# Local explanations are needed per instance, independent of any generation index.
# Generated deterministically (no API call) for the n=20 validity set.
print(f"Validity instances: {len(INSTANCE_IDS)}")
print(f"Loss option:   {LOSS_KEY}")
print(f"Output:        {EXPLANATIONS_DIR}")

Validity instances: 10
Loss option:   poisson_log
Output:        /Users/anton/Desktop/SoSe26/Belegarbeit/Implementation-XAI-Stahl-ss26/explanations


## 1. Load data and models

In [2]:
X_train, y_train, X_test, y_test = load_train_test()
print(f"X_train: {X_train.shape}  |  X_test: {X_test.shape}")

from utils.models import load_models
xgb, ebm = load_models(LOSS_KEY)
print("Models loaded.")

metrics = json.loads((RESULTS_DIR / f"model_metrics_{LOSS_KEY}.json").read_text())["metrics"]
print("Metrics loaded:", list(metrics.keys()))

X_train: (12152, 9)  |  X_test: (5227, 9)
Models loaded.
Metrics loaded: ['xgb', 'ebm']


## 2. Global explanations

In [3]:
print("Computing global XGBoost explanation (SHAP over the training set) ...")
global_xgb = build_global(xgb, "xgb", X_train, metrics["xgb"])

path = save_explanation(global_xgb, f"global_xgb_{LOSS_KEY}.json")
print(f"Saved: {path}")

print("\nTop 5 features (XGBoost):")
for entry in global_xgb["global_importance"][:5]:
    print(f"  #{entry['rank']:2d}  {entry['feature']:12s}  {entry['importance']:.4f}")

Computing global XGBoost explanation (SHAP over the training set) ...
Saved: /Users/anton/Desktop/SoSe26/Belegarbeit/Implementation-XAI-Stahl-ss26/explanations/global_xgb_poisson_log.json

Top 5 features (XGBoost):
  # 1  hr            0.8326
  # 2  yr            0.2004
  # 3  temp          0.1971
  # 4  weekday       0.1728
  # 5  mnth          0.1183


In [4]:
print("Computing global EBM explanation ...")
global_ebm = build_global(ebm, "ebm", X_train, metrics["ebm"])

path = save_explanation(global_ebm, f"global_ebm_{LOSS_KEY}.json")
print(f"Saved: {path}")

print("\nTop 5 features (EBM):")
for entry in global_ebm["global_importance"][:5]:
    print(f"  #{entry['rank']:2d}  {entry['feature']:12s}  {entry['importance']:.4f}")

Computing global EBM explanation ...
Saved: /Users/anton/Desktop/SoSe26/Belegarbeit/Implementation-XAI-Stahl-ss26/explanations/global_ebm_poisson_log.json

Top 5 features (EBM):
  # 1  hr            0.8888
  # 2  temp          0.2526
  # 3  yr            0.2166
  # 4  mnth          0.0857
  # 5  hum           0.0826


## 2b. Global shape / dependence curves + plots

Per feature: a curve JSON (`global_curve_{model}_{feature}.json`) plus a PNG.
EBM shape functions from `explain_global()`, XGB SHAP dependence + beeswarm.
Deterministic, no API call. Plots go to `explanations/plots/global/`.

In [5]:
# Global shape (EBM) and dependence (XGB) curves + plots. Deterministic, no API.
import matplotlib.pyplot as plt
import shap

PLOTS_GLOBAL_DIR = EXPLANATIONS_DIR / "plots" / "global"
PLOTS_GLOBAL_DIR.mkdir(parents=True, exist_ok=True)

# --- EBM: shape functions (bin midpoints / categories) ---
ebm_curves = build_global_curves(ebm, "ebm", X_train)
save_global_curves(ebm_curves, "ebm")


def _plot_ebm_shape(curve):
    fig, ax = plt.subplots(figsize=(6, 3.5))
    if curve["kind"] == "continuous":
        ax.plot(curve["x"], curve["y"], drawstyle="steps-mid", color="#1f77b4")
    else:
        ax.bar([str(c) for c in curve["x"]], curve["y"], color="#1f77b4")
    ax.axhline(0, color="grey", lw=0.8)
    ax.set_title(f"EBM shape: {curve['feature']} (imp={curve['importance']:.3f})")
    ax.set_xlabel(curve["feature"])
    ax.set_ylabel("contribution (log space)")
    fig.savefig(PLOTS_GLOBAL_DIR / f"ebm_shape_{curve['feature']}.png",
                dpi=130, bbox_inches="tight")
    plt.close(fig)


for c in ebm_curves:
    _plot_ebm_shape(c)
print(f"EBM: {len(ebm_curves)} shape curves + plots saved.")

# XGB: SHAP dependence (scatter) + beeswarm
xgb_curves = build_global_curves(xgb, "xgb", X_train)
save_global_curves(xgb_curves, "xgb")

explainer = shap.TreeExplainer(xgb)
sv = explainer(X_train)                       # full Explanation = beeswarm raw matrix
for feat in X_train.columns:
    shap.plots.scatter(sv[:, feat], show=False)
    plt.savefig(PLOTS_GLOBAL_DIR / f"xgb_dependence_{feat}.png",
                dpi=130, bbox_inches="tight")
    plt.close("all")
shap.plots.beeswarm(sv, show=False)
plt.savefig(PLOTS_GLOBAL_DIR / "xgb_beeswarm.png", dpi=130, bbox_inches="tight")
plt.close("all")
print(f"XGB: {len(xgb_curves)} dependence curves + "
      f"{len(X_train.columns)} scatter plots + beeswarm saved.")
print(f"Global artefacts in {EXPLANATIONS_DIR} and {PLOTS_GLOBAL_DIR}")


EBM: 9 shape curves + plots saved.
XGB: 9 dependence curves + 9 scatter plots + beeswarm saved.
Global artefacts in /Users/anton/Desktop/SoSe26/Belegarbeit/Implementation-XAI-Stahl-ss26/explanations and /Users/anton/Desktop/SoSe26/Belegarbeit/Implementation-XAI-Stahl-ss26/explanations/plots/global


## 2c. Beeswarm of EBM term contributions (SHAP analogue) + a-priori validation

The EBM's per-instance term contributions form a beeswarm directly (importance +
direction) - the analogue of the SHAP beeswarm. Note the split: the *contributions*
are native/exact (each is the global shape curve `f_j` evaluated at the instance,
straight from `eval_terms`, with no post-hoc approximation), whereas the beeswarm
*plot* is a SHAP-style visualisation rebuilt here from them. XGB, by contrast, needs
TreeSHAP - a post-hoc attribution - for its beeswarm.

**A-priori validation rule** (fixed *before* inspecting the numbers, to avoid
post-hoc rationalisation of the framing):

- **Primary (confound-free): validates the method.** The beeswarm feature order
  (mean |contribution|) must equal the EBM `explain_global` importance order, and
  each feature's colouring direction must be consistent with its global shape curve.
  Same model + same data -> no model-vs-method confound. If this holds, the EBM
  beeswarm is valid independently of XGB.
- **Secondary (with confound): confirmation only.** Concordance with the XGB SHAP
  beeswarm (Spearman of the feature order + sign agreement of the colouring). High
  concordance = both models tell the same story; low concordance ≠ method failure
  (may be a genuine model difference).

**Decision rule:** primary passed -> valid standalone contribution; secondary high ->
additionally report as cross-model confirmation, secondary low -> report as a
standalone EBM analysis (not as an error).

In [6]:
# Beeswarm of EBM term contributions + two-layer validation. Deterministic, no API.
import numpy as np
from utils.global_explain import (
    ebm_local_contributions, feature_importance_order, coloring_directions,
    beeswarm_plot, order_concordance, direction_agreement,
)

ebm_contrib = ebm_local_contributions(ebm, X_train)
beeswarm_plot(ebm_contrib, X_train, PLOTS_GLOBAL_DIR / "ebm_beeswarm.png")

# Primary validation (confound-free): beeswarm order vs EBM explain_global order.
ebm_order = feature_importance_order(ebm_contrib)
eg_order = [e["feature"] for e in global_ebm["global_importance"]]
prim = order_concordance(ebm_order, eg_order)
print(f"PRIMARY  (EBM beeswarm vs explain_global): "
      f"Spearman={prim['spearman']:.3f}, top3={prim['top3_overlap']}/3")

# Secondary concordance (with model confound): EBM vs XGB SHAP.
ebm_dirs = coloring_directions(ebm_contrib, X_train)
xgb_order = [c["feature"] for c in sorted(xgb_curves, key=lambda c: -c["importance"])]
xgb_dirs = {c["feature"]: int(np.sign(np.corrcoef(c["x"], c["y"])[0, 1]))
            for c in xgb_curves}
sec = order_concordance(ebm_order, xgb_order)
dir_ag = direction_agreement(ebm_dirs, xgb_dirs)
print(f"SECONDARY (EBM vs XGB SHAP): order Spearman={sec['spearman']:.3f}, "
      f"top3={sec['top3_overlap']}/3, direction agreement={dir_ag['agreement']:.2f} "
      f"({dir_ag['n_compared']} features)")
print(f"Saved: {PLOTS_GLOBAL_DIR / 'ebm_beeswarm.png'}")


PRIMARY  (EBM beeswarm vs explain_global): Spearman=1.000, top3=3/3
SECONDARY (EBM vs XGB SHAP): order Spearman=0.883, top3=3/3, direction agreement=1.00 (9 features)
Saved: /Users/anton/Desktop/SoSe26/Belegarbeit/Implementation-XAI-Stahl-ss26/explanations/plots/global/ebm_beeswarm.png


### A-priori evaluation

- **Primary — passed.** Beeswarm order == `explain_global` order (Spearman = 1.000,
  top-3 3/3); colouring directions consistent with the G0 shape curves (0 mismatches).
  → The beeswarm rests on the EBM's exact, native term contributions and is valid
  independently of XGB.
- **Secondary — high concordance.** EBM vs XGB SHAP: order Spearman = 0.883
  (top-3 3/3), colouring-direction agreement = 100 % (0/9 disagreements).

**Framing (per the a-priori rule):** primary passed → valid standalone contribution;
secondary high → additionally reported as cross-model confirmation.

*(The code cell above prints these numbers live; values recorded here from the
n=full-train run on 01.07.)*

## 3. Local explanations for the test instances

In [7]:
for n_done, iid in enumerate(INSTANCE_IDS, 1):
    local = build_local(xgb, "xgb", X_test, y_test, iid)
    path = save_explanation(local, f"local_xgb_{LOSS_KEY}_inst{iid}.json")
    if n_done <= 5 or n_done % 25 == 0 or iid in INSTANCE_IDS:
        top = local["contributions"][0]
        print(f"  [{n_done:3d}/{len(INSTANCE_IDS)}] inst={iid:4d}  pred={local['prediction']:6.1f}  "
              f"y_true={local['y_true']:6.0f}  top={top['feature']} ({top['contribution']:+.3f})")
print(f"XGB: {len(INSTANCE_IDS)} local explanations saved.")

  [  1/10] inst= 224  pred= 112.0  y_true=    96  top=temp (-0.549)
  [  2/10] inst= 580  pred=  64.6  y_true=    63  top=hr (-0.365)
  [  3/10] inst=1041  pred= 385.4  y_true=   387  top=hr (+0.746)
  [  4/10] inst=1481  pred= 228.6  y_true=   277  top=hr (+0.378)
  [  5/10] inst=1677  pred= 254.3  y_true=   286  top=hr (+0.252)
  [  6/10] inst=2058  pred= 194.1  y_true=   243  top=yr (-0.195)
  [  7/10] inst=2510  pred= 326.0  y_true=   372  top=hr (+0.686)
  [  8/10] inst=3543  pred= 309.9  y_true=   286  top=yr (+0.186)
  [  9/10] inst=3847  pred= 585.8  y_true=   531  top=hr (+0.413)
  [ 10/10] inst=4454  pred= 297.3  y_true=   354  top=hr (+0.352)
XGB: 10 local explanations saved.


In [8]:
for n_done, iid in enumerate(INSTANCE_IDS, 1):
    local = build_local(ebm, "ebm", X_test, y_test, iid)
    path = save_explanation(local, f"local_ebm_{LOSS_KEY}_inst{iid}.json")
    if n_done <= 5 or n_done % 25 == 0 or iid in INSTANCE_IDS:
        top = local["contributions"][0]
        print(f"  [{n_done:3d}/{len(INSTANCE_IDS)}] inst={iid:4d}  pred={local['prediction']:6.1f}  "
              f"y_true={local['y_true']:6.0f}  top={top['feature']} ({top['contribution']:+.3f})")
print(f"EBM: {len(INSTANCE_IDS)} local explanations saved.")

  [  1/10] inst= 224  pred= 117.0  y_true=    96  top=hr (+0.852)
  [  2/10] inst= 580  pred=  73.1  y_true=    63  top=yr (-0.226)
  [  3/10] inst=1041  pred= 389.7  y_true=   387  top=hr (+1.109)
  [  4/10] inst=1481  pred= 241.8  y_true=   277  top=hr (+0.819)
  [  5/10] inst=1677  pred= 260.2  y_true=   286  top=hr (+0.555)
  [  6/10] inst=2058  pred= 176.9  y_true=   243  top=hr (+0.555)
  [  7/10] inst=2510  pred= 311.4  y_true=   372  top=hr (+1.148)
  [  8/10] inst=3543  pred= 296.0  y_true=   286  top=hr (+0.555)
  [  9/10] inst=3847  pred= 593.4  y_true=   531  top=hr (+0.661)
  [ 10/10] inst=4454  pred= 331.8  y_true=   354  top=hr (+0.852)
EBM: 10 local explanations saved.


## 4. Overview of saved files

In [9]:
files = sorted(EXPLANATIONS_DIR.glob(f"*_{LOSS_KEY}*.json"))
print(f"Saved explanation files ({len(files)}):")
for f in files:
    size_kb = f.stat().st_size / 1024
    print(f"  {f.name:<45s}  {size_kb:.1f} KB")

Saved explanation files (22):
  global_ebm_poisson_log.json                    2.7 KB
  global_xgb_poisson_log.json                    2.7 KB
  local_ebm_poisson_log_inst1041.json            1.2 KB
  local_ebm_poisson_log_inst1481.json            1.2 KB
  local_ebm_poisson_log_inst1677.json            1.2 KB
  local_ebm_poisson_log_inst2058.json            1.2 KB
  local_ebm_poisson_log_inst224.json             1.1 KB
  local_ebm_poisson_log_inst2510.json            1.2 KB
  local_ebm_poisson_log_inst3543.json            1.2 KB
  local_ebm_poisson_log_inst3847.json            1.2 KB
  local_ebm_poisson_log_inst4454.json            1.2 KB
  local_ebm_poisson_log_inst580.json             1.2 KB
  local_xgb_poisson_log_inst1041.json            1.1 KB
  local_xgb_poisson_log_inst1481.json            1.2 KB
  local_xgb_poisson_log_inst1677.json            1.2 KB
  local_xgb_poisson_log_inst2058.json            1.2 KB
  local_xgb_poisson_log_inst224.json             1.1 KB
  local_xgb_poisso

## 5. Sanity check: XGBoost vs EBM top features

In [10]:
imp_xgb = {e["feature"]: e["rank"] for e in global_xgb["global_importance"]}
imp_ebm = {e["feature"]: e["rank"] for e in global_ebm["global_importance"]}

all_features = sorted(set(imp_xgb) | set(imp_ebm))
rows = [{"Feature": f, "Rank XGB": imp_xgb.get(f, "na"), "Rank EBM": imp_ebm.get(f, "na")} for f in all_features]
df = pd.DataFrame(rows).sort_values("Rank XGB")
print(df.to_string(index=False))

   Feature  Rank XGB  Rank EBM
        hr         1         1
        yr         2         3
      temp         3         2
   weekday         4         7
      mnth         5         4
       hum         6         5
weathersit         7         6
 windspeed         8         8
   holiday         9         9
